sdk/include/sdk/DataPointValues.h

```diff
@@ -0,0 +1,256 @@
+/**
+ * @brief See
+ * https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/timestamp.proto
+ */
+struct Timestamp {
+    int64_t seconds{0};
+    int32_t nanos{0};
+};
+
+inline bool operator==(const Timestamp& lhs, const Timestamp& rhs) {
+    return lhs.seconds == rhs.seconds && lhs.nanos == rhs.nanos;
+}
+
+class DataPoint;
+
+enum class DataPointFailure {
+    // No failure, i.e. is a valid data point
+    NONE,
+    // The data point is known, but doesn't have a valid value
+    INVALID_VALUE,
+    // The data point is known, but no value is available
+    NOT_AVAILABLE,
+    // Unknown datapoint
+    UNKNOWN_DATAPOINT,
+    // Access denied
+    ACCESS_DENIED,
+    // Unexpected internal error
+    INTERNAL_ERROR,
+};
+
+std::string toString(DataPointFailure);
+
+class DataPointValue {
+public:
+    enum class Type {
+        INVALID,
+        BOOL,
+        BOOL_ARRAY,
+        INT32,
+        INT32_ARRAY,
+        INT64,
+        INT64_ARRAY,
+        UINT32,
+        UINT32_ARRAY,
+        UINT64,
+        UINT64_ARRAY,
+        FLOAT,
+        FLOAT_ARRAY,
+        DOUBLE,
+        DOUBLE_ARRAY,
+        STRING,
+        STRING_ARRAY
+    };
+
+    DataPointValue(Type type, std::string path, Timestamp timestamp,
+                   DataPointFailure failure = DataPointFailure::NONE)
+        : m_path(std::move(path))
+        , m_type{type}
+        , m_timestamp(std::move(timestamp))
+        , m_failure{failure} {}
+
+    virtual ~DataPointValue()                        = default;
+    DataPointValue(const DataPointValue&)            = default;
+    DataPointValue(DataPointValue&&)                 = default;
+    DataPointValue& operator=(const DataPointValue&) = default;
+    DataPointValue& operator=(DataPointValue&&)      = default;
+
+    [[nodiscard]] const std::string& getPath() const { return m_path; }
+    [[nodiscard]] Type               getType() const { return m_type; }
+    [[nodiscard]] const Timestamp&   getTimestamp() const { return m_timestamp; }
+    [[nodiscard]] bool               isValid() const { return m_failure == DataPointFailure::NONE; }
+    [[nodiscard]] DataPointFailure   getFailure() const { return m_failure; }
+
+    bool operator==(const DataPointValue& other) const {
+        return std::tie(m_path, m_type, m_timestamp, m_failure) ==
+               std::tie(other.m_path, other.m_type, other.m_timestamp, other.m_failure);
+    }
+
+private:
+    std::string      m_path;
+    Type             m_type{Type::INVALID};
+    Timestamp        m_timestamp{};
+    DataPointFailure m_failure{DataPointFailure::NONE};
+};
+
+template <typename T> DataPointValue::Type getValueType() {
+    static_assert(std::is_same<T, std::false_type>::value, "Value type not supported!");
+    return DataPointValue::Type::INVALID;
+}
+
+template <> inline DataPointValue::Type getValueType<bool>() { return DataPointValue::Type::BOOL; }
+template <> inline DataPointValue::Type getValueType<std::vector<bool>>() {
```

Comment:
Maybe we should typedef the std::vector<T> to something to be able to replace vector with a different implementation easily.

Reply:
I'd prefer doings this in a later refactoring